radix-engine-interface/src/model/resource/non_fungible_local_id.rs

```diff
@@ -70,55 +70,71 @@ impl From<u128> for NonFungibleLocalId {
 /// Represents the local id of a non-fungible.
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub enum NonFungibleLocalId {
+    /// String of `[_0-9a-zA-Z]{1,64}`.
     String(String),
-    Number(u64),
+    /// Unsigned integers, up to u64.
+    Integer(u64),
+    /// Bytes, of length between 1 and 64.
     Bytes(Vec<u8>),
+    /// UUID, v4, variant 1, big endian. See https://www.rfc-editor.org/rfc/rfc4122
     UUID(u128),
 }
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub enum IdValidationError {
+pub enum ContentValidationError {
     TooLong,
     Empty,
     ContainsBadCharacter(char),
+    NotUuidV4Variant1,
 }
 
 impl NonFungibleLocalId {
     pub fn id_type(&self) -> NonFungibleIdType {
         match self {
             NonFungibleLocalId::String(..) => NonFungibleIdType::String,
-            NonFungibleLocalId::Number(..) => NonFungibleIdType::Number,
+            NonFungibleLocalId::Integer(..) => NonFungibleIdType::Integer,
             NonFungibleLocalId::Bytes(..) => NonFungibleIdType::Bytes,
             NonFungibleLocalId::UUID(..) => NonFungibleIdType::UUID,
         }
     }
 
-    pub fn validate_contents(&self) -> Result<(), IdValidationError> {
+    pub fn validate_contents(&self) -> Result<(), ContentValidationError> {
```

Comment:
I actually have a note to myself to go back to NFs and ensure we add a validate in all sensible places (eg when extracting from a manifest, or in the schema checker) - but I think it'd be best to ensure this with the compiler.

Reply:
Feels like over engineering, essentially to represent different states of non-fungible id: `ValidatedNonFungibleLocalId` and `NonFungibleLocalId`. 

Do we really think clients need such protection? or is it making construction cumbersome?

Reply:
I'm just aware that I missed some cases where I didn't validate the ID in my original PR before Christmas, and I imagine there might well be others (eg sys-calls / decode impl / schema validation) where it might be easy to miss it too.

I don't think we'd want a `ValidatedNonFungibleLocalId`, I think we'd probably want `NonFungibleLocalId` (the struct that people see) and `NonFungibleLocalIdContent` (the enum which they'd rarely need to see), where constructing a `NonFungibleLocalId` takes the content (or has methods for each id kind) and ensures we validate it after construction.

Reply:
I'd suggest we reconsider it at another time. The `IntoNonFungibleLocalId` (to be renamed) trait family heavily depends on error-free non-fungible id construction.